lib/MongoDB/_Protocol.pm

```diff
@@ -77,6 +79,126 @@ use constant {
 #     substr( $msg, 0, 4, pack( P_INT32, length($msg) ) );
 
+use constant {
+    # length for MsgHeader
+    P_HEADER_LENGTH =>
+        length(pack P_HEADER, 0, 0, 0, 0),
+    # length for OP_COMPRESSED
+    P_COMPRESSED_PREFIX_LENGTH =>
+        length(pack P_COMPRESSED, 0, 0, 0, 0, 0, 0, 0),
+};
+
+# struct OP_COMPRESSED {
+#     MsgHeader header;             // standard message header
+#     int32_t   originalOpcode;     // wrapped op code
+#     int32_t   uncompressedSize;   // size of deflated wo. header
+#     uint8_t   compressorId;       // compressor
+#     char*     compressedMessage;  // compressed contents
+# };
+
+# compressor dispatchers
+#   $name => [
+#       $compressor_id,
+#       $compress,          # $compress->($buf, $opts)
+#       $decompress,        # $decompress->($buf)
+#   ],
+my %COMPRESSOR = (
+    none => [
+        0,
+        sub { shift },
+        sub { shift },
+    ],
+    zlib => [
+        2,
+        sub {
+            my ($buf, $opts) = @_;
+
+            my $level = $opts->{zlib_compression_level};
+            $level = undef
+                if defined $level and $level < 0;
+
+            require Compress::Zlib;
+            return Compress::Zlib::compress(
+                $buf,
+                defined($level) ? $level : (),
+            );
+        },
+        sub {
+            require Compress::Zlib;
+            return Compress::Zlib::uncompress(shift);
+        },
+    ],
+);
+
+# determines if an op code should be compressed
+sub is_compressible {
+    my ($msg) = @_;
+
+    my ($len, $request_id, $response_to, $op_code)
+        = unpack(P_HEADER, $msg);
+
+    return grep { $_ == $op_code }
+        OP_QUERY,
+        OP_MSG,
+        OP_REPLY,
+        OP_INSERT,
+        OP_UPDATE,
+        OP_DELETE,
+        OP_GET_MORE,
+        OP_KILL_CURSORS;
+}
+
+# compress message
+sub compress {
+    my ($msg, $compressor, %opts) = @_;
+
+    my ($comp_id, $comp_cb) = @{ $COMPRESSOR{$compressor} || [] }
+        or MongoDB::ProtocolError->throw("Unknown compressor '$compressor'");
+
+    my ($len, $request_id, $response_to, $op_code)
+        = unpack(P_HEADER, $msg);
+
+    $msg = substr $msg, P_HEADER_LENGTH;
+
+    my $msg_comp = pack(
+        P_COMPRESSED,
+        0, $request_id, $response_to, OP_COMPRESSED,
+        $op_code,
+        length($msg),
+        $comp_id,
+    ).$comp_cb->($msg, \%opts);
+
+    substr($msg_comp, 0, 4, pack(P_INT32, length($msg_comp)));
+    return $msg_comp;
+}
+
+# attempt to uncompress message
+# messages that aren't OP_COMPRESSED are returned as-is
+sub try_uncompress {
+    my ($msg) = @_;
+
+    my ($len, $request_id, $response_to, $op_code, $orig_op_code, undef, $comp_id)
+        = unpack(P_COMPRESSED, $msg);
+
+    return $msg
+        if $op_code != OP_COMPRESSED;
+
+    $msg = substr $msg, P_COMPRESSED_PREFIX_LENGTH;
+
+    for my $comp_name (keys %COMPRESSOR) {
```

Comment:
Rather than looping (O(N)), just build a file-scoped decompressor array that you can index into:

`my @DECOMPRESSORS = [ sub { shift }, undef, \&_decompress_zlib ];`

Reply:
And if the compressor becomes a coderef attribute of server, then you don't need the %COMPRESSOR hash at all, only an array of decompressor coderefs.